geode-core/src/main/java/org/apache/geode/distributed/internal/InternalLocator.java

```diff
@@ -1043,7 +1046,15 @@ public void waitToStop() throws InterruptedException {
         while (system.isConnected()) {
           Thread.sleep(5000);
         }
+        // there would be a gap between stoppedForReconnect being to true and attemptingToReconnect
+        // being true, if system.waitUntilReconnected happened in between, this method would return
+        // with "restarted" being false, so we need also to wait till system is reconnecting
```

Comment:
I kind of understand what's going on. Could this comment be rephrased so that instead of talking about private fields of `InternalDistributedSystem`, it talked about the contract provided by the `DistributedSystem` interface?

Also I wonder about the other two places in production code where we call `IDS.waitUntilReconnected()`. How do those two places avoid the same problem we see here (before this PR)?

The other two places are: `InternalLocator.attemptReconnect()` and `GemFireCacheImpl.waitUntilReconnected()`. Neither of those call `IDS.isReconnecting()` so it makes me think that this method should follow suit. If not, I think the comment should explain why we're doing it differently here.

Reply:
`InternalLocator.attemptReconnect()` doesn't care about the return boolean, for `GemFireCacheImpl.waitUntilReconnected()`, I am not sure how it's using the return boolean. I need this `IDS.isReconnecting()` here specifically because `locator.waitToStop()` as the method name indicate, can't return unless the locator truly stops. So the call to the `IDS.waitUntilReconnected()` can't return false.